File: p057h/insert_interval.js
/**
 * Definition for an interval.
 */
 function Interval(start, end) {
    this.start = start;
    this.end = end;
 }

/**
 * @param {Interval[]} intervals
 * @param {Interval} newInterval
 * @return {Interval[]}
 */
var insert = function(intervals, newInterval) {
    let ilen = intervals.length;
    let result = [];
    let i = 0;
    while (i < ilen && intervals[i].end < newInterval.start) {
        result.push(intervals[i++]);
    }
    let mergedStart = newInterval.start, mergedEnd = newInterval.end;
    while (i < ilen && intervals[i].start <= newInterval.end) {
        mergedStart = Math.min(intervals[i].start, mergedStart);
        mergedEnd   = Math.max(intervals[i].end, mergedEnd);
        i++;
    }
    result.push(new Interval(mergedStart, mergedEnd));
    while (i < ilen) {
        result.push(intervals[i++]);
    }
    return result;
};
// TEST
[
    {
        intervals: [],
        newInterval: [2,5]
    },
    {
        intervals: [[1,3]],
        newInterval: [4,5]
    },
    {
        intervals: [[4,8]],
        newInterval: [2,3]
    },
    {
        intervals: [[4,8]],
        newInterval: [4,7]
    },
    {
        intervals: [[4,8]],
        newInterval: [3,9]
    },
    {
        intervals: [[1,3],[6,9]],
        newInterval: [2,5]
    },
    {
        intervals: [[1,2],[3,5],[6,7],[8,10],[12,16]],
        newInterval: [4,9]
    },
].forEach(function (test) {
    let intervals = [];
    test.intervals.forEach(function (elem) {
        intervals.push(new Interval(elem[0], elem[1]));
    });
    let newInterval = new Interval(test.newInterval[0], test.newInterval[1]);
    console.log("Insert", test.newInterval, "into", test.intervals, "->");
    let merged = insert(intervals, newInterval).map(function (i) {
        return [i.start, i.end];
    });
    console.log(merged);
});

File: p057h/insert_interval.py
from typing import List


class Solution:
    def insertV1(
        self, intervals: List[List[int]], newInterval: List[int]
    ) -> List[List[int]]:
        ans, inserted = [], False
        ns, ne = newInterval
        for s, e in intervals:
            if e < ns:
                ans.append([s, e])
            elif s > ne:
                if not inserted:
                    ans.append([ns, ne])
                    inserted = True
                ans.append([s, e])
            else:
                ns = min(ns, s)
                ne = max(ne, e)

        if not inserted:
            ans.append([ns, ne])
            inserted = True
        return ans

    def insertV2(
        self, intervals: List[List[int]], newInterval: List[int]
    ) -> List[List[int]]:
        left, right = [], []
        ns, ne = newInterval
        for s, e in intervals:
            if e < ns:
                left.append([s, e])
            elif s > ne:
                right.append([s, e])
            else:
                ns = min(ns, s)
                ne = max(ne, e)

        return left + [[ns, ne]] + right


# TESTS
tests = [
    ([], [2, 5], [[2, 5]]),
    ([[1, 3]], [4, 5], [[1, 3], [4, 5]]),
    ([[4, 8]], [2, 3], [[2, 3], [4, 8]]),
    ([[4, 8]], [4, 7], [[4, 8]]),
    ([[4, 8]], [3, 9], [[3, 9]]),
    ([[1, 3], [6, 9]], [2, 5], [[1, 5], [6, 9]]),
    ([[1, 2], [3, 5], [6, 7], [8, 10], [12, 16]], [4, 8], [[1, 2], [3, 10], [12, 16]]),
]
for intervals, new, expected in tests:
    sol = Solution()
    actual = sol.insertV1(intervals, new)
    print("Insert", new, "into", intervals, "->", actual)
    assert actual == expected
    assert expected == sol.insertV2(intervals, new)
